docs/ProcedureExecution.md

```diff
@@ -108,7 +108,45 @@ struct StorageCapReadOutput {
 
 ```
 
-When reading, the parameters are parsed as a list of u256 keys, and for each key a value is returned. When writing, we parse two lists of keys and values to be written. Output in this case are the old values being written over. 
+When reading, the parameters are parsed as a list of u256 keys, and for each key a value is returned. When writing, we parse two lists of keys and values to be written. Output in this case are the old values being written over.
+
+#### Summary of Message Format
+
+The format of the memory message is as follows: 1 byte for the capability type, one byte for the capability index. Each procedure has a list of capabilities, and the capability index is an index into this list. (**TODO**: Why do we need to specify the capability type here? Isn't that contained within the capability?). After that is the start address where we will write, as well as the number of addresses we will write to.
+
+* Byte 0: Capability Type
+* Byte 1: Capability Index
+* Bytes 2 to 34: Number of addresses to be written to (this values is referred to as `n`)
+* Bytes 34 to 34+n\*3`: Storage keys to be written to.
+* Bytes 34+n\*3 to 34+2\*n\*32: Values to be written into each corresponding key.
+
+For example, to write the value `0x34` into the storage key `0x7` the message format will be as follows.
+
+* Byte 0: `0x7`
+* Byte 1: A single byte corresponding to the index of this capability in the list.
+* Bytes 2-33: `0x1`, we are writing to a single address.
```

Comment:
If we don't use padding, we can save space by removing the length of the payload. (Just split the u256 values by two, you get keys and values)

Reply:
In that case we could even strip them k1,v1,k2,v2, which would make the kernel code simpler at the expense of the contract. It would be an interesting cost comparison.